// amplify-category-function@2.34.7/lib/provider-utils/awscloudformation/utils/updateDependentFunctionCfn.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateDependentFunctionsCfn = void 0;
const amplify_cli_core_1 = require("amplify-cli-core");
const execPermissionsWalkthrough_1 = require("../service-walkthroughs/execPermissionsWalkthrough");
const lambda_walkthrough_1 = require("../service-walkthroughs/lambda-walkthrough");
const loadFunctionParameters_1 = require("./loadFunctionParameters");
const path = __importStar(require("path"));
const constants_1 = require("./constants");
const constants_2 = require("../../../constants");
async function updateDependentFunctionsCfn(context, dependentFunctionResource, backendDir, modelsDeleted, apiResource) {
    for (const lambda of dependentFunctionResource) {
        const resourceDirPath = path.join(backendDir, constants_2.categoryName, lambda.resourceName);
        const currentParameters = loadFunctionParameters_1.loadFunctionParameters(resourceDirPath);
        const selectedCategories = currentParameters.permissions;
        let categoryPolicies = [];
        let permissions = {};
        let resources = [];
        const functionParameters = {
            resourceName: lambda.resourceName,
            environmentMap: {
                ENV: {
                    Ref: 'env',
                },
                REGION: {
                    Ref: 'AWS::Region',
                },
            },
        };
        for (const selectedCategory of Object.keys(selectedCategories)) {
            const selectedResources = selectedCategories[selectedCategory];
            for (const resourceName of Object.keys(selectedResources)) {
                if (!modelsDeleted.includes(resourceName)) {
                    const resourcePolicy = selectedResources[resourceName];
                    const { permissionPolicies, cfnResources } = await execPermissionsWalkthrough_1.getResourcesForCfn(context, resourceName, resourcePolicy, apiResource, selectedCategory);
                    categoryPolicies = categoryPolicies.concat(permissionPolicies);
                    if (!permissions[selectedCategory]) {
                        permissions[selectedCategory] = {};
                    }
                    permissions[selectedCategory][resourceName] = resourcePolicy;
                    resources = resources.concat(cfnResources);
                }
            }
        }
        const { environmentMap, dependsOn } = await execPermissionsWalkthrough_1.generateEnvVariablesForCfn(context, resources, {});
        functionParameters.categoryPolicies = categoryPolicies;
        functionParameters.mutableParametersState = { permissions };
        functionParameters.environmentMap = environmentMap;
        functionParameters.dependsOn = dependsOn;
        functionParameters.lambdaLayers = currentParameters.lambdaLayers;
        lambda_walkthrough_1.updateCFNFileForResourcePermissions(resourceDirPath, functionParameters, currentParameters, apiResource);
        currentParameters.permissions = permissions;
        const parametersFilePath = path.join(resourceDirPath, constants_1.functionParametersFileName);
        amplify_cli_core_1.JSONUtilities.writeJson(parametersFilePath, currentParameters);
        lambda.dependsOn = functionParameters.dependsOn;
        context.amplify.updateamplifyMetaAfterResourceUpdate(constants_2.categoryName, lambda.resourceName, 'dependsOn', lambda.dependsOn);
    }
}
exports.updateDependentFunctionsCfn = updateDependentFunctionsCfn;
