// amplify-cli-core@1.29.0/lib/hooks/hooksConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.skipHooksFilePath = exports.defaultSupportedExt = exports.supportedEnvEvents = exports.suppportedEvents = exports.hookFileSeperator = void 0;
exports.hookFileSeperator = '-';
exports.suppportedEvents = {
    add: [
        'notifications',
        'analytics',
        'api',
        'auth',
        'function',
        'hosting',
        'interactions',
        'predictions',
        'storage',
        'xr',
        'codegen',
        'env',
    ],
    update: ['notifications', 'analytics', 'api', 'auth', 'function', 'hosting', 'interactions', 'predictions', 'storage', 'xr', 'env'],
    remove: ['notifications', 'analytics', 'api', 'auth', 'function', 'hosting', 'interactions', 'predictions', 'storage', 'xr', 'env'],
    push: ['analytics', 'api', 'auth', 'function', 'hosting', 'interactions', 'storage', 'xr'],
    pull: ['env'],
    publish: [],
    delete: [],
    checkout: ['env'],
    list: ['env'],
    get: ['env'],
    mock: ['api', 'storage', 'function'],
    build: ['function'],
    status: ['notifications'],
    import: ['auth', 'storage', 'env'],
    gqlcompile: ['api'],
    addgraphqldatasource: ['api'],
    statements: ['codegen'],
    types: ['codegen'],
};
exports.supportedEnvEvents = ['add', 'update', 'remove', 'pull', 'checkout', 'list', 'get', 'import'];
exports.defaultSupportedExt = { js: { runtime: 'node' }, sh: { runtime: 'bash' } };
exports.skipHooksFilePath = '/opt/amazon';
